clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go

```diff
@@ -0,0 +1,280 @@
+	// nodeSyncEventReason is the event reason emitted by service controller when it completes node sync on the lb.
+	nodeSyncEventReason = "UpdatedLoadBalancer"
+
+	phaseNodeSyncStart    = "nodesync_triggered"
+	phaseNodeSyncCompelte = "nodesync_complete"
+)
+
+// Scheme is the default instance of runtime.Scheme to which types in the Kubernetes API are already registered.
+// This is needed for ConfigMap search.
+var Scheme = runtime.NewScheme()
+
+var nodeSyncTransition = map[string]measurementutil.Transition{
+	"nodesync_start_to_complete": {
+		From: phaseNodeSyncStart,
+		To:   phaseNodeSyncCompelte,
+	},
+}
+
+func init() {
+	if err := measurement.Register(loadBalancerNodeSyncLatencyName, createLoadBalancerNodeSyncMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", loadBalancerNodeSyncLatencyName, err)
+	}
+
+	// Register external types for Scheme
+	v1.AddToScheme(Scheme)
+}
+
+func createLoadBalancerNodeSyncMeasurement() measurement.Measurement {
+	return &LoadBalancerNodeSyncMeasurement{
+		selector:                  measurementutil.NewObjectSelector(),
+		svcNodeSyncLatencyTracker: measurementutil.NewObjectTransitionTimes(loadBalancerNodeSyncLatencyName),
+	}
+}
+
+type LoadBalancerNodeSyncMeasurement struct {
+	client clientset.Interface
+	// selector used to select relevant load balancer type service used for measurement
+	selector *measurementutil.ObjectSelector
+	// waitTimeout specify for the timeout for node sync on all LBs to complete
+	waitTimeout time.Duration
+	// svcNodeSyncLatencyTracker tracks the nodesync latency
+	svcNodeSyncLatencyTracker *measurementutil.ObjectTransitionTimes
+	// triggerNodeName is the node name which used to trigger LB nodesync
+	triggerNodeName string
+	// triggerTimestamp is the timestamp when nodesync is trigger
+	triggerTimestamp time.Time
+}
+
+// LoadBalancerNodeSyncMeasurement takes measurement of node sync latency for selected lb type services.
+// This measurement only works for K8s 1.19 as it depends on the ExcludeNodeForLoadbalancer label.
+// Services can be specified by field and/or label selectors.
+// If namespace is not passed by parameter, all LoadBalancer type service with all-namespace scope is assumed.
+// "measure" action triggers nodesync and observation of nodesync completion for selected LB services.
+// "gather" returns node sync latency summary.
+func (s *LoadBalancerNodeSyncMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	s.client = config.ClusterFramework.GetClientSets().GetClient()
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	switch action {
+	case "measure":
+		if err := s.selector.Parse(config.Params); err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultNodeSyncLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.measureNodeSyncLatency()
```

Comment:
The measure step picks one candidate node, marks it to be excluded from LB and then measures the time taken for the update event to show up. 

This is a good test, but I think the most problematic scenarios are where all nodes are removed and replaced by a new set of nodes, causing discrepancy in the k8s and GCE layer until steady state. Can we trigger that scenario by doing kubectl delete of all nodes?

Reply:
The problem with that is kubelet will try to add the node back immediately. The reason why I did not pick a rolling update or node pool operation is because it might cause multiple nodesyncs. Causing the unnecessary noise for measuring latency.